### ISSUE TITLE
'dotnet publish' with --runtime argument won't produce stand alone app

### ISSUE THREAD

**Issue by billwert** [dotnet/cli#2103]
I am informed (for now) that the way to build a stand alone app (opposed to shared library) is to add runtimes to your project.json and to invoke 'dotnet publish' with --runtime choosing one of them. This does not seem to work on today's CLI.
Repro:
1) dotnet new
2) notepad project.json, add runtimes tag like so:
```  
"runtimes": {
	"win7-x64": {},
	"win7-x86": {}
  },
```
3) dotnet restore
4) dotnet build
5) dotnet publish -r win7-x64

Result:
C:\source\test3>dotnet publish -r win7-x64
'C:\source\test3' cannot be published for '<no framework provided>' 'win7-x64'

Adding -f netstandard1.5 doesn't help.

The resultant output of publishing without -r is: 
```
 Directory of C:\source\test3\bin\Debug\netstandard1.5\publish

03/28/2016  07:49 PM    <DIR>          .
03/28/2016  07:49 PM    <DIR>          ..
03/28/2016  07:57 PM                 0 test3.deps
03/28/2016  07:57 PM            83,051 test3.deps.json
03/28/2016  07:57 PM             4,608 test3.dll
03/28/2016  07:57 PM            11,776 test3.pdb
03/28/2016  07:57 PM               135 test3.runtimeconfig.json
               5 File(s)         99,570 bytes
```

I did notice that dotnet --info says my RID is win10-x64, but using that instead doesn't help.

C:\source\test3>dotnet --info
.NET Command Line Tools (1.0.0-beta-002071)

Product Information:
 Version:     1.0.0-beta-002071
 Commit Sha:  cf616770ca

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.10586
 OS Platform: Windows
 RID:         win10-x64

**Comment by joepeer** [comment]
Instead of netstandard1.5 put netstandardapp1.5 in the framework section of your project.json

Not sure why the template for dotnet new  type console is set to netstandard1.5.

With the runtimes section also added both dotnet restore and publish should work.

**Comment by billwert** [comment]
netstandardapp1.5 doesn't fix it.

**Comment by blackdwarf** [comment]
@anurse @piotrMSFT can you take a look at this? Seems like a bug.

**Comment by brthor** [comment]
@joepeer my understanding is dotnet new template is set to netstandard1.5 as a Point in Time until RID inference is removed in NuGet for netcoreapp1.0

**Comment by pakrym** [comment]
"type":"platform" should be removed from dependencies

**Comment by billwert** [comment]
@pakrym perhaps, but that doesn't help this issue. Tried it and no change.

**Comment by piotrpMSFT** [comment]
@pakrym @anurse you guys collaborated fairly closely on the standalone publish scenario. Can you please help get @billwert unblocked?

**Comment by billwert** [comment]
@anurse was kind enough to help me figure this out offline.

**Comment by piotrpMSFT** [comment]
@billwert can you share the solution so other folks have access?

**Comment by hultqvist** [comment]
@piotrpMSFT I managed to get an exe following this
https://github.com/dotnet/core/issues/77#issuecomment-219807733

**Comment by stajs** [comment]
Since @billwert hasn't responded with the solution, perhaps @anurse would be so kind?

**Comment by blackdwarf** [comment]
@stajs is the problem still reproing for you? If so, can you provide more details? Thanks!

**Comment by GuardRex** [comment]
@stajs If you can shoot your `project.json` in here, we can take a look. I only ever use self-contained apps (unless I'm forced go portable when held at :gun:point).

Generally, it's just going with `"Microsoft.NETCore.App": "1.0.0"` in deps [no `"type": "platform"` ... and don't comment it out as shown in the linked issue (comments are not supported in `project.json`)], a `netcoreapp1.0` framework, adding the runtimes section with your RID of choice (e.g., `win10-x64`), and publishing with `dotnet publish -r <RID> -c Release` (e.g., `dotnet publish -r win10-x64 -c Release`). That should be all it takes.

**Comment by blackdwarf** [comment]
@GuardRex yep, it is the same way for me, really.

**Comment by stajs** [comment]
My scenario is probably worthy of a separate issue. I'm depending on full .NET Framework 4.6.1, because I'm installing a Windows Service that self-hosts an ASP.NET Core website.

```json
// project.json for Windows Service project
"frameworks": {
	"net461": {
		"frameworkAssemblies": {
			"System.ServiceProcess": "4.0.0.0",
			"System.Runtime": "4.0.20.0"
		}
	}
},
```

Development is on a mixture of Windows 7/8.1 machines and being published to Windows 10. Based on discussions in this and other issues, I've tried adding a RID to the publish target like so:

```json
// project.json for Windows Service project
"runtimes": {
	"win10-x64": {}
},
```

I've tried both `x86` and `x64` (I have no particular dependence on bitness). The publish command is run from the Windows Service project:

```
dotnet publish --configuration Release --framework net461 --runtime win10-x64
```

Output is showing:

```
Publishing <Redacted>.Service for .NETFramework,Version=v4.6.1/win10-x64
```

My version is:

```
H:\>dotnet --version
1.0.0-preview2-003121
```

Installing on the Windows 10 publish target (which does not have .NET Core SDK) starts the Windows Service, which then starts the self-hosted ASP.NET Core website. Turned on the developer middleware (`app.UseDeveloperExceptionPage()`) and see the following in the browser:

> InvalidOperationException: Can not find reference assembly '.NETFramework/v4.6.1/Microsoft.CSharp.dll' file for package Microsoft.CSharp

![image](https://cloud.githubusercontent.com/assets/2253814/18455506/67eaa0a8-799e-11e6-82d5-2c7f96b08706.png)

According to MSDN, .NET Framework 4.6.1 is installed on the publish target:

https://msdn.microsoft.com/en-us/library/hh925568(v=vs.110).aspx

![image](https://cloud.githubusercontent.com/assets/2253814/18455546/be6fce94-799e-11e6-9ee0-b70b83ba2c8b.png)

![image](https://cloud.githubusercontent.com/assets/2253814/18455629/4003bb78-799f-11e6-8079-dbfa75dbc529.png)

Previously we were using DNX RC1, and when published all the dependencies ended up in `approot\packages\<adependencies>`. Now it looks like the new way is just a flat list of DLLs - is that correct, or should I be expecting more "packages"?

**Comment by blackdwarf** [comment]
@stajs 

> Now it looks like the new way is just a flat list of DLLs - is that correct, or should I be expecting more "packages"?

Yes, this is correct, the self-contained option is a flat list of DLLs.

**Comment by stajs** [comment]
I'm having a hard time wrapping my head around the "other option".

[According to the docs](https://docs.microsoft.com/en-us/dotnet/articles/core/deploying/index), the deploy options are Framework-dependent deployments (FDD) or Self-contained deployments (SCD). "Framework" is an overloaded term and lends to my confusion, but I take it that FDD means that your publish target needs to have the same version of .NET Core installed as the builder. I am trying to use SCD because:

0. The publish target does not have .NET Core installed.
0. I want the published code to use the version of .NET Core at the time of building.

SCD in my case is an oxymoron because I still need full .NET Framework dependencies. I'm happy to leave that part of my scenario aside for now because I don't think that is the cause of my problem, i.e. it looks like SCD doesn't include all of .NET Core?

So to compare, firstly with a `win10-x64` RID in the `project.json` for the Windows Service, and running this command you can see it starts to publish with the same RID:

```
C:\<path>\<redacted>.Service>dotnet publish --configuration Release --runtime win10-x64 --build-base-path C:\<path>\build\temp --output C:\<path>\build\temp\publish-with-RID
Publishing Saber.Shed.Service for .NETFramework,Version=v4.6.1/win10-x64
```

After removing the RID from both the `project.json` and the command, it defaults to `win7-x64` on my Windows 8.1 machine:

```
C:\<path>\<redacted>.Service>dotnet publish --configuration Release --build-base-path C:\<path>\build\temp --output C:\<path>\build\temp\publish-without-RID
Publishing Saber.Shed.Service for .NETFramework,Version=v4.6.1/win7-x64
```

I output the above commands to two different directories, and a diff shows some differences but only for DLLs for my own application code.

![image](https://cloud.githubusercontent.com/assets/2253814/18496151/b24791ea-7a77-11e6-973c-b9a6e053230a.png)

![image](https://cloud.githubusercontent.com/assets/2253814/18496248/978aa44a-7a78-11e6-9e35-f5d83721b8dc.png)

Since I don't see an appreciable difference, I have to assume that I'm doing it wrong.

**Comment by svick** [comment]
@stajs If you're compiling for `.NETFramework,Version=v4.6.1`, then you're not targeting .Net Core at all, so I don't think the distinction between the two deploy options applies to you.

**Comment by stajs** [comment]
@svick That makes sense. I expected there to be _some_ difference in the output, so perhaps a wrong expectation on my part.

I think I've found the problem that was blocking me, I'll update once I've confirmed...

**Comment by GuardRex** [comment]
The difference will be that the second command falls back to the runtime for `win7-x64`, while the former explicitly names `win10-x64`. They are forward-compat per a convo I had with @ericstj ages ago.

... and you can definitely have a .NET 4.6 Framework app this is an ASP.NET Core 1.0 app. From Scott's blog at http://www.hanselman.com/blog/ASPNET5IsDeadIntroducingASPNETCore10AndNETCore10.aspx ...

![diagram](https://cloud.githubusercontent.com/assets/1622880/18497233/122b648c-79f2-11e6-8b3f-15ef8b2b406e.png)

**Comment by stajs** [comment]
I take it back:

> SCD in my case is an oxymoron because I still need full .NET Framework dependencies. I'm happy to leave that part of my scenario aside for now because I don't think that is the cause of my problem, i.e. it looks like SCD doesn't include all of .NET Core?

Looking again at the exception:

> InvalidOperationException: Can not find reference assembly '.NETFramework/v4.6.1/Microsoft.CSharp.dll' file for package Microsoft.CSharp

The clue being "reference assembly":

![image](https://cloud.githubusercontent.com/assets/2253814/18498882/55beb374-7a8e-11e6-85f1-9e4267ce9c84.png)

The publish target is missing a `v4.6.1`:

![image](https://cloud.githubusercontent.com/assets/2253814/18497416/2b982e56-7a82-11e6-9c1a-5e3898269575.png)

If I manually take a copy of my reference assemblies over to the publish target, everything works.

---

I shouldn't have trusted the registry. I think that the publish target must have had the right version of full .NET Framework at some stage for the flag to be set, but I can only assume that it has been removed since then. Checking _Programs and Features_ shows only a single item identifying with .NET Framework:

![image](https://cloud.githubusercontent.com/assets/2253814/18498446/994e3ae6-7a8a-11e6-8602-fcce7aaf475e.png)

**Comment by stajs** [comment]
Actually, it seems those reference assemblies are just for [design time](http://stackoverflow.com/questions/8153048/missing-referenced-assemblies-folder-for-net-4-0) [support](http://www.hanselman.com/blog/NETVersioningAndMultiTargetingNET45IsAnInplaceUpgradeToNET40.aspx):

![image](https://cloud.githubusercontent.com/assets/2253814/18499669/14dcc330-7a95-11e6-9528-4e8e3ecdd065.png)

I need to figure out why the published DLLs are trying to load the reference assemblies. Is it a :bug:? It seems like it should be getting them from the GAC.

I'm guessing my machine has the reference assemblies because of Visual Studio or other SDKs. I've just verified that installing [4.6.2](https://www.microsoft.com/en-us/download/details.aspx?id=53344) doesn't add reference assemblies by itself.

**Comment by stajs** [comment]
I'm unblocked now, see https://github.com/dotnet/cli/issues/4177#issuecomment-247187268.
